salt/states/file.py

```diff
@@ -557,7 +557,18 @@ def process_symlinks(filenames, symlinks):
             managed_directories.add(mdest)
             keep.add(mdest)
 
-    return managed_files, managed_directories, managed_symlinks, keep
+    # Sets are randomly ordered. We need to make sure symlinks are always at the
+    # end, so we need to use a list
```

Comment:
Creating a new separate list here is fine but an alternative would be to use a `dict` instead of the set in the first place since dict in python is now ordered since python 3.7.

Reply:
From glancing over the existing code it appears that a set is being returned for all 3 managed_files, managed_directories, and managed_symlinks but then it is iterating over them which means that the children if a directory could always be getting visited first which would mean the `makedirs=True` is going to make the parent directories

```
 _ret = directory(
            path,
            user=user,
            group=group,
            recurse=[],
            dir_mode=dir_mode,
            file_mode=None,
            makedirs=True,
            clean=False,
            require=None,
        )
```

and then later on the mode/group user etc will be corrected. I don't think this is ideal. I think it would be ideal if we return a `list`/`dict.keys()` for all of them.

Reply:
Yes... I would prefer not to use sets at all in this case. The only reason I can think a set was used was to avoid duplicates. You're right, ``dict.keys()`` would also not allow duplicates.